### PR TITLE
Profiling Training Loop

### PR DESCRIPTION
**Pull request by indiravats**
Modified squad_dataloader to optionally select a subset of examples for debugging 
Added profiling for memory and flops using torch profiler (this should work on custom models built on top of the torch.nn.Module and torch operations)

TO DO: verify if profiling works for minGRU

### CODE REVIEW

File: train_squad.py [hunk 1]
```diff
@@ -7,6 +7,9 @@
 from transformers import get_cosine_schedule_with_warmup
 from tqdm import tqdm
 
+import psutil
```
Comment (dlrudwo1269) [inline train_squad.py:10]: Don't think this is used

File: train_squad.py [hunk 2]
```diff
@@ -41,6 +44,14 @@
...
 B = 16 # batch size
 
+"""
+# Testing configs
+max_lr = 1e-3 
+min_lr = max_lr * 0.1
+epochs = 10
+B = 10 
+"""
+
```
Comment (dlrudwo1269) [inline train_squad.py:54]: Can we clean these up before merging in

File: train_squad.py [hunk 3]
```diff
@@ -63,8 +74,15 @@
 
 model.to(device)
 
+# num examples could be set here if needed for debugging 
+"""
+NUM_EXAMPLES = 1000
+train_loader, val_loader = get_squad_v2_dataloaders(tokenizer, batch_size=B, num_examples=NUM_EXAMPLES)
+references = get_squad_v2_validation_references(num_examples=NUM_EXAMPLES)
+"""
```
Comment (dlrudwo1269) [inline train_squad.py:82]: ditto

File: train_squad.py [hunk 4]
```diff
@@ -79,9 +97,11 @@
 
 optimizer = torch.optim.AdamW(model.parameters(), lr=1e-4, fused=use_fused)
 num_training_steps = epochs * len(train_loader)  # Total number of steps
+"""
+warmup_steps = int(0.1 * num_training_steps)
+"""
```
Comment (dlrudwo1269) [inline train_squad.py:102]: ditto

File: train_squad.py [hunk 5]
```diff
@@ -167,17 +187,37 @@ def get_predictions(batch, logits):
...
-        loss_accum += loss.detach()
-        loss.backward()
-        optimizer.step()
-        scheduler.step()
+        with torch.profiler.profile(
+            activities=[torch.profiler.ProfilerActivity.CPU, torch.profiler.ProfilerActivity.CUDA],
+            record_shapes=False,
+            profile_memory=False,
+            with_flops=True,
+        ) as prof:
```
Comment (dlrudwo1269) [inline train_squad.py:208]: Is there a reason why a new profiler is set up every batch? Wouldn't this introduce overhead?

File: train_squad.py [hunk 6]
```diff
@@ -189,7 +229,18 @@ def get_predictions(batch, logits):
...
     dt = t1 - t0
     tok_per_sec = tokens_processed / dt
-    epoch_metrics = f"[Train] Epoch {i:4d} | loss: {avg_loss:.6f} | cur_lr: {cur_lr:.6f} | dt: {dt:.2f} | tok/sec: {tok_per_sec:.2f}"
+    flops_in_tera = total_flops / 1e12
+
+    # getting peak memory usage for the current epoch
+    if device == "cuda":
+        max_memory = torch.cuda.max_memory_allocated(device=device) / (1024 * 1024)  # convert to MB
+    else:
+        max_memory = process.memory_info().rss / (1024 * 1024)  # MB
```
Comment (dlrudwo1269) [inline train_squad.py:238]: I think you meant to use pstuils here. However, wouldn't this just get the cpu memory usage at this point in time after all the computation graph has been discarded, etc.?

File: src/utils/squad_dataloader.py [hunk 7]
```diff
@@ -104,8 +109,12 @@ def collate_fn(batch):
     return train_dataloader, val_dataloader
 
 
-def get_squad_v2_validation_references():
+def get_squad_v2_validation_references(num_examples=None):
     val_dataset = load_dataset("rajpurkar/squad_v2", split="validation")
+
+    if num_examples is not None:
+      val_dataset = val_dataset.select(range(num_examples))
+
```
Comment (dlrudwo1269) [inline src/utils/squad_dataloader.py:117]: Does this get the same set of examples for the validation dataset? Also we can probably use a better ratio for train vs validation size.

**Review by dlrudwo1269 — COMMENTED** [review]
General logic looks good but please do the sanity check before merging

File: train_squad.py (hunk 5 above)
Comment (indiravats) [reply train_squad.py:208]: This was done when I was reporting the batch-level metrics as well, thanks for pointing this out.

Comment (indiravats) [reply train_squad.py:208]: Just realized that changing this was what was giving me the 0 FLOPs error initally, looking into it.

File: src/utils/squad_dataloader.py (hunk 7 above)
Comment (indiravats) [reply src/utils/squad_dataloader.py:117]: the val dataloader has shuffle=False so the order of examples remains consistent; have fixed the ratio

File: train_squad.py (hunk 6 above)
Comment (indiravats) [reply train_squad.py:238]: fixed

File: train_squad.py [hunk 8]
```diff
@@ -167,18 +171,37 @@ def get_predictions(batch, logits):
...
+        for batch_idx, batch in enumerate(tqdm(train_loader)):
+            optimizer.zero_grad()
+            logits, loss = forward_batch(batch)
+            loss_accum += loss.detach()
+            loss.backward()
+            optimizer.step()
+            scheduler.step()
+            tokens_processed += batch["input_ids"].shape[0] * batch["input_ids"].shape[1] # batch_size * sequence_length
+
+    epoch_flops += sum(
```
Comment (dlrudwo1269) [inline train_squad.py:203]: Non-blocking: We can probably just assign here instead of += and remove `epoch_flops = 0` above

**Review by dlrudwo1269 — APPROVED** [review]
